### PR TITLE
Getting started with spring security and spring boot

### PR DESCRIPTION
**Pull request by ranjanih**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2022/2022-10-25-getting-started-with-spring-security.md [hunk 1]
```diff
@@ -0,0 +1,1011 @@
...
+Now, we should be able to hit the endpoint from postman without passing credentials:
+
+{{% image alt="settings" src="images/posts/getting-started-with-spring-security/unsecure.JPG" %}}
+
+#### 5. Add Custom Filters
+Spring provides security by executing a sequence of filters in a chain. In cases where we need to add additional checks to the request
+before it reaches the controller, Spring Security provides us with the below methods that help us add a custom filter at the desired position in the chain.
+- **addFilterBefore(Filter filter, Class<? extends Filter> beforeFilter)**: This method lets us add the custom filter before the specified filter in the chain.
+- **addFilterAfter(Filter filter, Class<? extends Filter> afterFilter)**: This method lets us add the custom filter after the specified filter in the chain.
+- **addFilterAt(Filter filter, Class<? extends Filter> atFilter)**: This method lets us add the custom filter at the specified filter in the chain.
```
Comment (arkuksin) [inline content/blog/2022/2022-10-25-getting-started-with-spring-security.md:577]: What does it mean? Is the old filter overridden at this location? If not, in which order will they be executed?

File: content/blog/2022/2022-10-25-getting-started-with-spring-security.md [hunk 2]
```diff
@@ -0,0 +1,1011 @@
...
+    }
+}
+````
+- **`@WithMockUser(username = "bookadmin", roles = {"USER"})`** : Here we are running the test with the username `bookadmin` and role `USER`. This test is used only to verify if the endpoint is secured.
+Further we have also used methods `authenticated()` to verify the authentication details and hamcrest matcher `hasSize()` to verify the response object.
+- **`@WithMockUser(username = "bookadmin", roles = {"ADMIN"})`** : Here, we get a Forbidden response since the roles do not match. Although the user is mocked, roles re required to match for a success response.
+- When no user details are specified, the endpoint is not secured and therefore we get Unauthorized response.
+
+### @WithUserDetails
+Instead of mocking the user, we could also use the `UserDetailsService` bean created in the configuration.
```
Comment (arkuksin) [inline content/blog/2022/2022-10-25-getting-started-with-spring-security.md:940]: you mean a test configuration?

**Review by arkuksin — CHANGES_REQUESTED** [review]

File: content/blog/2022/2022-10-25-getting-started-with-spring-security.md [hunk 3]
```diff
@@ -0,0 +1,1011 @@
...
+
+````java
+class SpringBootWebSecurityConfiguration {
+    @ConditionalOnDefaultWebSecurity
+    static class SecurityFilterChainConfiguration {
+        SecurityFilterChainConfiguration() {
+        }
+
+        @Bean
+        @Order(2147483642)
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-10-25-getting-started-with-spring-security.md:245]: This is the code snippet from the decompiled class version of "SpringBootWebSecurityConfiguration". Will reword it to reflect this info.

File: content/blog/2022/2022-10-25-getting-started-with-spring-security.md [hunk 4]
```diff
@@ -0,0 +1,1011 @@
...
+        SecurityFilterChain defaultSecurityFilterChain(HttpSecurity http) throws Exception {
+            ((AuthorizedUrl) http.authorizeRequests().anyRequest()).authenticated();
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-10-25-getting-started-with-spring-security.md:247]: This is the code snippet from the decompiled class version of "SpringBootWebSecurityConfiguration". Will reword it to reflect this info.
